### PR TITLE
add lsp code

### PR DESCRIPTION
**Pull request by stackoverflow**
This is the same code as the old Pkl PR, now in stand alone form.

### CODE REVIEW

**Review by bioball — CHANGES_REQUESTED** [review]
Did a first pass!

Also: I noticed that resolution of parameters in object bodies wasn't working. I dug into it, and this diff fixed it for me:

```diff
diff --git a/src/main/kotlin/org/pkl/lsp/ast/Member.kt b/src/main/kotlin/org/pkl/lsp/ast/Member.kt
index 827129e..437cd43 100644
--- a/src/main/kotlin/org/pkl/lsp/ast/Member.kt
+++ b/src/main/kotlin/org/pkl/lsp/ast/Member.kt
@@ -18,6 +18,7 @@ package org.pkl.lsp.ast
 import org.pkl.core.parser.antlr.PklParser.*
 import org.pkl.lsp.LSPUtil.firstInstanceOf
 import org.pkl.lsp.PklVisitor
+import org.pkl.lsp.VirtualFile
 
 class PklClassPropertyImpl(override val parent: Node, override val ctx: ClassPropertyContext) :
   AbstractNode(parent, ctx), PklClassProperty {
@@ -102,10 +103,63 @@ class PklParameterListImpl(override val parent: Node, override val ctx: Paramete
   }
 }
 
+/**
+ * Workaround: The Pkl parser in pkl-core represents an object body's parameters as simply a
+ * sequence of nodes, whereas the logic copied from pkl-intellij assumes that it is a
+ * [PklParameterList].
+ *
+ * This provides the glue from pkl-core's syntax tree to what we want here.
+ */
+class PklParameterListOfObjectBodyImpl(
+  override val parent: Node,
+  private val parameters: List<ParameterContext>,
+) : PklParameterList {
+  override val elements: List<PklParameter> by lazy {
+    parameters.map { PklParameterImpl(this, it) }
+  }
+
+  override val span: Span
+    get() {
+      val start =
+        children.firstOrNull()?.span
+          ?: return Span(
+            parent.span.beginCol,
+            parent.span.endLine,
+            parent.span.beginCol,
+            parent.span.endLine,
+          )
+      val end = children.last().span
+      return if (start == end) start
+      else Span(start.beginCol, start.beginCol, end.endLine, end.endCol)
+    }
+
+  override val children: List<Node> by lazy { elements }
+
+  override val containingFile: VirtualFile
+    get() = parent.containingFile
+
+  override val enclosingModule: PklModule?
+    get() = parent.enclosingModule
+
+  override val terminals: List<Terminal>
+    get() = emptyList()
+
+  override val text: String
+    get() = parameters.joinToString(", ") { it.text }
+
+  override fun <R> accept(visitor: PklVisitor<R>): R? {
+    return visitor.visitParameterList(this)
+  }
+}
+
 class PklObjectBodyImpl(override val parent: Node, override val ctx: ObjectBodyContext) :
   AbstractNode(parent, ctx), PklObjectBody {
-  override val parameterList: List<PklParameter> by lazy {
-    children.filterIsInstance<PklParameter>()
+  override val parameterList: PklParameterList by lazy {
+    PklParameterListOfObjectBodyImpl(this, ctx.parameter())
+  }
+
+  override val children: List<Node> by lazy {
+    listOf(parameterList) + super.children.filterNot { it is PklParameter }
   }
 
   override val members: List<PklObjectMember> by lazy {
diff --git a/src/main/kotlin/org/pkl/lsp/ast/Node.kt b/src/main/kotlin/org/pkl/lsp/ast/Node.kt
index c1daf44..a8c90b9 100644
--- a/src/main/kotlin/org/pkl/lsp/ast/Node.kt
+++ b/src/main/kotlin/org/pkl/lsp/ast/Node.kt
@@ -294,7 +294,7 @@ sealed interface PklObjectBodyOwner : Node {
 }
 
 interface PklObjectBody : Node {
-  val parameterList: List<PklParameter>
+  val parameterList: PklParameterList
 
   val members: List<PklObjectMember>
 
diff --git a/src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt b/src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt
index c28ac34..0ee9821 100644
--- a/src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt
+++ b/src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt
@@ -36,7 +36,7 @@ class HoverFeature(override val server: PklLSPServer) : Feature(server) {
       val line = params.position.line + 1
       val col = params.position.character + 1
       val hoverText =
-        mod.findBySpan(line, col)?.let { resolveHover(it, line, col) } ?: return Hover(listOf())
+        mod.findBySpan(line, col)?.let { resolveHover(it, line, col, it) } ?: return Hover(listOf())
       return Hover(MarkupContent("markdown", hoverText))
     }
     return server.builder().runningBuild(params.textDocument.uri).thenApply(::run)
@@ -246,7 +246,7 @@ class HoverFeature(override val server: PklLSPServer) : Feature(server) {
           val computedType =
             Resolvers.resolveUnqualifiedAccess(
               originalNode,
-              null,
+              node.computeThisType(PklBaseModule.instance, mapOf()),
               true,
               PklBaseModule.instance,
               mapOf(),
diff --git a/src/main/kotlin/org/pkl/lsp/resolvers/Resolvers.kt b/src/main/kotlin/org/pkl/lsp/resolvers/Resolvers.kt
index bccf841..b170a59 100644
--- a/src/main/kotlin/org/pkl/lsp/resolvers/Resolvers.kt
+++ b/src/main/kotlin/org/pkl/lsp/resolvers/Resolvers.kt
@@ -195,7 +195,7 @@ object Resolvers {
               if (
                 !visitor.visitIfNotNull(
                   parameter.typedIdentifier?.identifier?.text,
-                  parameter,
+                  parameter.typedIdentifier,
                   functionType.bindings,
                 )
               )
@@ -260,12 +260,12 @@ object Resolvers {
               if (!visitor.visitIfNotNull(member.name, member, bindings))
                 return visitor.result to LookupMode.LEXICAL
             }
-            element.parameterList.let { parameterList ->
+            element.parameterList.elements.let { parameterList ->
               for (parameter in parameterList) {
                 if (
                   !visitor.visitIfNotNull(
                     parameter.typedIdentifier?.identifier?.text,
-                    parameter,
+                    parameter.typedIdentifier,
                     bindings,
                   )
                 )
@@ -305,7 +305,7 @@ object Resolvers {
               if (
                 !visitor.visitIfNotNull(
                   parameter.typedIdentifier?.identifier?.text,
-                  parameter,
+                  parameter.typedIdentifier,
                   bindings,
                 )
               )
diff --git a/src/main/kotlin/org/pkl/lsp/type/ComputeDefinitionType.kt b/src/main/kotlin/org/pkl/lsp/type/ComputeDefinitionType.kt
index fde84d1..17b8f59 100644
--- a/src/main/kotlin/org/pkl/lsp/type/ComputeDefinitionType.kt
+++ b/src/main/kotlin/org/pkl/lsp/type/ComputeDefinitionType.kt
@@ -146,17 +146,18 @@ fun Node?.computeResolvedImportType(
                 }
               }
             }
-            is PklParameterList ->
-              when (val parameterListOwner = identifierOwner.parent) {
+            is PklParameter -> {
+              val parameterList = identifierOwner.parent as PklParameterList
+              when (val parameterListOwner = parameterList.parent) {
                 is PklFunctionLiteralExpr -> {
                   val functionType = parameterListOwner.inferExprTypeFromContext(base, bindings)
-                  getFunctionParameterType(this, identifierOwner, functionType, base)
+                  getFunctionParameterType(this, parameterList, functionType, base)
                 }
                 is PklObjectBody ->
                   when (val objectBodyOwner = parameterListOwner.parent) {
                     is PklExpr -> {
                       val functionType = objectBodyOwner.computeExprType(base, bindings)
-                      getFunctionParameterType(this, identifierOwner, functionType, base)
+                      getFunctionParameterType(this, parameterList, functionType, base)
                     }
                     is PklObjectBodyOwner -> {
                       @Suppress("BooleanLiteralArgument")
@@ -168,12 +169,13 @@ fun Node?.computeResolvedImportType(
                           false,
                           cache,
                         )
-                      getFunctionParameterType(this, identifierOwner, functionType, base)
+                      getFunctionParameterType(this, parameterList, functionType, base)
                     }
                     else -> Type.Unknown
                   }
                 else -> Type.Unknown
               }
+            }
             else -> Type.Unknown
           }
         }
```

This is the scenario that wasn't working:

```pkl
class Person {
    name: String
}

local foo: (Int) -> Person = () -> new {}

bar = (foo) { num ->
//            ^ <caret>

}
```

Also: this initial implementation has no tests. We should start adding them, although we don't need to block this PR on it.

File: src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+      val line = params.position.line + 1
+      val col = params.position.character + 1
+      val hoverText =
+        mod.findBySpan(line, col)?.let { resolveHover(it, line, col) } ?: return Hover(listOf())
+      return Hover(MarkupContent("markdown", hoverText))
+    }
+    return server.builder().runningBuild(params.textDocument.uri).thenApply(::run)
+  }
+
+  private fun resolveHover(node: Node, line: Int, col: Int, originalNode: Node? = null): String? {
```
Comment (bioball) [inline src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt:45]: FYI: the way that this works on the pkl-intellij side is that `node` is the _resolved_ value (probably by resolving the node if it is a `PsiReference`), and `originalNode` is the element that is underneath the cursor.

If you've copied over the logic in pkl-intellij, you'll want to copy the same behavior in `resolveHover`.

File: src/main/kotlin/org/pkl/lsp/CacheManager.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import java.nio.file.Path
+import kotlin.io.path.*
+import org.pkl.lsp.ast.PklModule
+
+/** Manages all Pkl files that are not local to the file system: http(s), packages. */
+object CacheManager {
+  val pklDir = Path.of(System.getProperty("user.home")).resolve(".pkl")
+  val pklCacheDir = pklDir.resolve(".cache")
+
+  val lspCacheDir = Files.createTempDirectory("pkl-lsp-cache")
```
[the remark this answers is not in the document]
Comment (stackoverflow) [reply src/main/kotlin/org/pkl/lsp/CacheManager.kt:30]: Totally agree. But I think we can do that in a separate PR.

File: src/main/kotlin/org/pkl/lsp/analyzers/ModifierAnalyzer.kt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    private val TYPE_ALIAS_MODIFIERS = setOf(EXTERNAL, LOCAL)
+    private val CLASS_METHOD_MODIFIERS = setOf(ABSTRACT, EXTERNAL, LOCAL, CONST)
+    private val CLASS_PROPERTY_MODIFIERS = setOf(ABSTRACT, EXTERNAL, HIDDEN, LOCAL, FIXED, CONST)
+    private val OBJECT_METHOD_MODIFIERS = setOf(LOCAL)
+    private val OBJECT_PROPERTY_MODIFIERS = setOf(LOCAL)
+  }
+
+  override fun doAnalyze(node: Node, diagnosticsHolder: MutableList<PklDiagnostic>): Boolean {
+    if (node !is ModifierListOwner || node.modifiers == null) {
+      return node is PklModule || node is PklClass || node is PklClassBody
```
[the remark this answers is not in the document]
Comment (stackoverflow) [reply src/main/kotlin/org/pkl/lsp/analyzers/ModifierAnalyzer.kt]: Fixed

File: src/main/kotlin/org/pkl/lsp/features/Feature.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    return when {
+      prop.type != null -> prop
+      prop.isLocal -> prop
+      else -> {
+        val receiverType = prop.computeThisType(base, mapOf())
+        Resolvers.resolveQualifiedAccess(receiverType, true, base, visitor)
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (stackoverflow) [reply src/main/kotlin/org/pkl/lsp/features/Feature.kt]: You are right. I moved these functions to a new Node interface (PklReference) which is more in line with how our IntelliJ plugin works.

File: src/main/kotlin/org/pkl/lsp/PklTextDocumentService.kt [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  override fun definition(
+    params: DefinitionParams
+  ): CompletableFuture<Either<MutableList<out Location>, MutableList<out LocationLink>>> {
+    return definition.onGoToDefinition(params)
+  }
+
+  override fun completion(
+    params: CompletionParams
+  ): CompletableFuture<Either<MutableList<CompletionItem>, CompletionList>> {
```
[the remark this answers is not in the document]
Comment (stackoverflow) [reply src/main/kotlin/org/pkl/lsp/PklTextDocumentService.kt]: Indeed. Changed them to use `List`.

File: src/main/kotlin/org/pkl/lsp/NodeVisitor.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.pkl.lsp.ast.Node
+
+abstract class NodeVisitor {
+  companion object {
+    val EMPTY_VISITOR: NodeVisitor = object : NodeVisitor() {}
+  }
+
+  fun visitInnerElement(element: Node) {}
+}
```
[the remark this answers is not in the document]
Comment (stackoverflow) [reply src/main/kotlin/org/pkl/lsp/NodeVisitor.kt]: Yeah, this class was not used for anything. Removed.

File: src/main/kotlin/org/pkl/lsp/VirtualFile.kt [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  fun toModule(): PklModule?
+
+  companion object {
+    fun fromUri(uri: URI, logger: ClientLogger): VirtualFile? {
+      return if (uri.scheme.equals("file", ignoreCase = true)) {
+        val file = Paths.get(uri).toFile()
+        if (!file.isFile || file.extension != "pkl") {
+          return null
+        }
+        FsFile(file)
```
[the remark this answers is not in the document]
Comment (stackoverflow) [reply src/main/kotlin/org/pkl/lsp/VirtualFile.kt]: This was some kind of sanity check, but it's indeed better to let clients be in charge of what constitutes a Pkl file. Our vscode plugin already understands `PklProject` as a Pkl file. So nothing has to be done there.

File: src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt (hunk 1 above)
Comment (stackoverflow) [reply src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt]: Added a `resolveReference` used by the hover and go to definition.

**Review by bioball — APPROVED** [review]
Let's merge it!

**Pull request merged**